#### Module import

In [101]:
import pandas as pd
import os
import numpy as np
from sklearn.model_selection import train_test_split # Data Split
import statsmodels.api as sm

In [2]:
os.getcwd()

'c:\\Users\\Owner\\Desktop\\윤태준\\소'

##### Path

In [3]:
os.chdir('./3차 요청자료/소_3대 정리')
os.getcwd()

'c:\\Users\\Owner\\Desktop\\윤태준\\소\\3차 요청자료\\소_3대 정리'

##### Check
> 또 마지막 변수명 외증조로 중복돼있음, 일단 파일 한번에 불러서 합친 다음 변수명 다시 바꾸기

In [ ]:
# 확인
a1 = pd.read_excel('근내지방3_0.xlsx')
a1.head()

In [5]:
a1.columns

Index(['이표번호', 'KPN', '성별', '출생일', '농가코드', '농가등급', '도축개월', '도축일', '등급', '도체중',
       '등심단면적', '등지방두께', '근내지방도', '육색', '지방색', '성숙도', '조직도', '어미 이표번호',
       '형매 도체중 평균', '형매 등심단면적 평균', '형매 등지방 평균', '형매 근내지방 평균', '형매 마릿수',
       '외할미 이표번호', '어미형매 도체중 평균', '어미형매 등심단면적 평균', '어미형매 등지방 평균',
       '어미형매 근내지방 평균', '어미형매 마릿수', '외증조할미 이표번호', '외할미형매 도체중 평균',
       '외할미형매 등심단면적 평균', '외할미형매 등지방 평균', '외할미형매 근내지방 평균', '외할미형매 마릿수',
       '외고조할미 이표번호', '외증조할미형매 도체중 평균', '외증조할미형매 등심단면적 평균', '외증조할미형매 등지방 평균',
       '외증조할미형매 근내지방 평균', '외증조할미형매 마릿수', '외고고조할미 이표번호', '외증조할미형매 도체중 평균.1',
       '외증조할미형매 등심단면적 평균.1', '외증조할미형매 등지방 평균.1', '외증조할미형매 근내지방 평균.1',
       '외증조할미형매 마릿수.1'],
      dtype='object')

In [6]:
# 파일이 존재하는 경로 불러오기 및 리스트
path = os.getcwd()
file_lst = os.listdir(path)# 파일 이름 담기
file_lst

['근내지방3_0.xlsx',
 '근내지방3_1.xlsx',
 '근내지방3_2.xlsx',
 '근내지방3_3.xlsx',
 '근내지방3_4.xlsx',
 '근내지방3_5.xlsx',
 '근내지방3_6.xlsx',
 '근내지방3_7.xlsx']

##### ↓ 반복문으로 파일 불러오면서, 하나의 DataFrame으로 합침 

In [7]:
df_3 = pd.DataFrame()
for i in file_lst:
    data_3 = pd.read_excel(path +'/'+i)
    df_3 = pd.concat([df_3, data_3])
    
df_3 = df_3.reset_index(drop = True)

In [8]:
# 확인
print(df_3.shape)
df_3.head()

(110345, 47)


이표번호      KPN  성별         출생일    농가코드 농가등급  도축개월         도축일    등급  \
0  2137173111  KPN1053   암  2019-03-28  398841  1+C    42  2022-09-21    1B   
1  2150109802  KPN1060  거세  2020-04-28  263150   1A    28  2022-08-22  1++C   
2  2133835356  KPN1099   암  2018-09-28  189147   1A    52  2023-01-03    1A   
3  2156452289  KPN1284   암  2020-11-18  451221   1A    27  2023-01-26    1A   
4  2147835365  KPN1060   암  2019-10-23  190620   1A    32  2022-06-07    2A   

   도체중  ...  외증조할미형매 등심단면적 평균  외증조할미형매 등지방 평균  외증조할미형매 근내지방 평균  외증조할미형매 마릿수  \
0  374  ...             89.00           16.33             2.33            3   
1  524  ...             83.33           13.67             6.00            3   
2  333  ...              0.00            0.00             0.00            0   
3  372  ...              0.00            0.00             0.00            0   
4  378  ...              0.00            0.00             0.00            0   

   외고고조할미 이표번호  외증조할미형매 도체중 평균.1  외증조할미형매 등심단면적 평균.1  외증조할미형매 등지방 평균.1  \
0  198856328.0               0.0                 0.0               0.0   
1  172418108.0               0.0                 0.0               0.0   
2          NaN               0.0                 0.0               0.0   
3          NaN               0.0                 0.0               0.0   
4          NaN               0.0                 0.0               0.0   

   외증조할미형매 근내지방 평균.1  외증조할미형매 마릿수.1  
0                0.0              0  
1                0.0              0  
2                0.0              0  
3                0.0              0  
4                0.0              0  

[5 rows x 47 columns]

##### 변수명 재설정

In [9]:
df_3.columns

Index(['이표번호', 'KPN', '성별', '출생일', '농가코드', '농가등급', '도축개월', '도축일', '등급', '도체중',
       '등심단면적', '등지방두께', '근내지방도', '육색', '지방색', '성숙도', '조직도', '어미 이표번호',
       '형매 도체중 평균', '형매 등심단면적 평균', '형매 등지방 평균', '형매 근내지방 평균', '형매 마릿수',
       '외할미 이표번호', '어미형매 도체중 평균', '어미형매 등심단면적 평균', '어미형매 등지방 평균',
       '어미형매 근내지방 평균', '어미형매 마릿수', '외증조할미 이표번호', '외할미형매 도체중 평균',
       '외할미형매 등심단면적 평균', '외할미형매 등지방 평균', '외할미형매 근내지방 평균', '외할미형매 마릿수',
       '외고조할미 이표번호', '외증조할미형매 도체중 평균', '외증조할미형매 등심단면적 평균', '외증조할미형매 등지방 평균',
       '외증조할미형매 근내지방 평균', '외증조할미형매 마릿수', '외고고조할미 이표번호', '외증조할미형매 도체중 평균.1',
       '외증조할미형매 등심단면적 평균.1', '외증조할미형매 등지방 평균.1', '외증조할미형매 근내지방 평균.1',
       '외증조할미형매 마릿수.1'],
      dtype='object')

In [10]:
# 원래 변수명
df_3.columns = ['이표번호', 'KPN', '성별', '출생일', '농가코드', '농가등급', '도축개월', '도축일', '등급', '도체중',
       '등심단면적', '등지방두께', '근내지방도', '육색', '지방색', '성숙도', '조직도', '어미 이표번호',
       '형매 도체중 평균', '형매 등심단면적 평균', '형매 등지방 평균', '형매 근내지방 평균', '형매 마릿수',
       '외할미 이표번호', '어미형매 도체중 평균', '어미형매 등심단면적 평균', '어미형매 등지방 평균',
       '어미형매 근내지방 평균', '어미형매 마릿수', '외증조할미 이표번호', '외할미형매 도체중 평균',
       '외할미형매 등심단면적 평균', '외할미형매 등지방 평균', '외할미형매 근내지방 평균', '외할미형매 마릿수',
       '외고조할미 이표번호', '외증조할미형매 도체중 평균', '외증조할미형매 등심단면적 평균', '외증조할미형매 등지방 평균',
       '외증조할미형매 근내지방 평균', '외증조할미형매 마릿수', '외고고조할미 이표번호', '외고조할미형매_도체중_평균', '외고조할미형매_등심단면적_평균', '외고조할미형매_등지방_평균', '외고조할미형매_근내지방_평균',
    '외고조할미형매_마릿수'
]

In [11]:
cow_3 = df_3.copy()
cow_3 = cow_3[['이표번호', 'KPN', '성별', '출생일', '농가코드', '농가등급', '도축개월', '도축일', '등급', '도체중',
       '등심단면적', '등지방두께', '근내지방도', '육색', '지방색', '성숙도', '조직도', '어미 이표번호',
       '형매 도체중 평균', '형매 등심단면적 평균', '형매 등지방 평균', '형매 근내지방 평균', '형매 마릿수',
       '외할미 이표번호', '어미형매 도체중 평균', '어미형매 등심단면적 평균', '어미형매 등지방 평균',
       '어미형매 근내지방 평균', '어미형매 마릿수', '외증조할미 이표번호', '외할미형매 도체중 평균',
       '외할미형매 등심단면적 평균', '외할미형매 등지방 평균', '외할미형매 근내지방 평균', '외할미형매 마릿수']]

In [13]:
print(cow_3.shape)
cow_3.columns

(110345, 35)


Index(['이표번호', 'KPN', '성별', '출생일', '농가코드', '농가등급', '도축개월', '도축일', '등급', '도체중',
       '등심단면적', '등지방두께', '근내지방도', '육색', '지방색', '성숙도', '조직도', '어미 이표번호',
       '형매 도체중 평균', '형매 등심단면적 평균', '형매 등지방 평균', '형매 근내지방 평균', '형매 마릿수',
       '외할미 이표번호', '어미형매 도체중 평균', '어미형매 등심단면적 평균', '어미형매 등지방 평균',
       '어미형매 근내지방 평균', '어미형매 마릿수', '외증조할미 이표번호', '외할미형매 도체중 평균',
       '외할미형매 등심단면적 평균', '외할미형매 등지방 평균', '외할미형매 근내지방 평균', '외할미형매 마릿수'],
      dtype='object')

In [14]:
# 원 데이터와 변수명 변경 후 데이터 차이 확인
# numpy의 setdiff1d(a, b)를 이용해서, a와 b를 비교하여 b에 없는 것 확인
print(f'변경 후 변수명 : {np.setdiff1d(df_3.columns, cow_3.columns)}')
print(f'변경 전 변수명 : {np.setdiff1d(cow_3.columns, df_3.columns)}')

변경 후 변수명 : ['외고고조할미 이표번호' '외고조할미 이표번호' '외고조할미형매_근내지방_평균' '외고조할미형매_도체중_평균'
 '외고조할미형매_등심단면적_평균' '외고조할미형매_등지방_평균' '외고조할미형매_마릿수' '외증조할미형매 근내지방 평균'
 '외증조할미형매 도체중 평균' '외증조할미형매 등심단면적 평균' '외증조할미형매 등지방 평균' '외증조할미형매 마릿수']
변경 전 변수명 : []


In [15]:
cow_3.성별.value_counts()

거세    68328
암     42017
Name: 성별, dtype: int64

##### 성별 및 근내지방도 변수변환
> <p> <span style = 'color : orange' > 성별 : {F:암, N_M:거세} <span> </p>
> <p> <span style = 'color : orange' > 근내지방도 : {1 : (7~9), 0: (0~6)} <span> </p>

In [16]:
# 성별 암, 거세 => F, N_M 변환
cow_3.성별 = cow_3.성별.apply(lambda x: 'F' if x == '암' else 'N_M' if x =='거세' else x )
print('변경 전 성별 unique 및 counts')
print(df_3.성별.value_counts())
print('\n')
print('변경 후 성별 unique 및 counts')
print(cow_3.성별.value_counts())

변경 전 성별 unique 및 counts
거세    68328
암     42017
Name: 성별, dtype: int64


변경 후 성별 unique 및 counts
N_M    68328
F      42017
Name: 성별, dtype: int64


In [20]:
# 근내지방도 0, 1 변경
cow_3['근내지방도'] = cow_3['근내지방도'].apply(lambda x: 1 if x >= 7 else 0 if x < 7 else x)
print('변경 전 근내지방도 unique 및 counts')
print(df_3.근내지방도.value_counts())
print('\n')
print('변경 후 근내지방도 unique 및 counts')
print(cow_3.근내지방도.value_counts())

변경 전 근내지방도 unique 및 counts
6.0    27779
4.0    17824
5.0    13561
7.0    12600
8.0    10126
3.0     9495
9.0     8984
2.0     7891
1.0     1911
Name: 근내지방도, dtype: int64


변경 후 근내지방도 unique 및 counts
0.0    78461
1.0    31710
Name: 근내지방도, dtype: int64


##### full data export

In [21]:
cow_3.to_csv('../소_3대 정리/3_total_utf.csv', index=False)
cow_3.to_csv('../소_3대 정리/3_total_cp.csv',encoding= 'cp949', index=False)

<h5> train_test_split, 7:3 </h5>

In [28]:
X_3 = cow_3.drop(['근내지방도'],axis=1)
Y_3 = cow_3.근내지방도
X_train_3, X_test_3, Y_train_3, Y_test_3 = train_test_split(X_3, Y_3, test_size=0.3, random_state = 2023)

<h5> 값 다 같은 인덱스, 원래 같은 행이므로 이대로 합쳐도 원래 데이터 셋이 됨 </h5>

In [24]:
train_3 = pd.concat([X_train_3,Y_train_3], axis=1)
test_3 = pd.concat([X_test_3,Y_test_3], axis=1)

In [25]:
print(train_3.shape)
print(test_3.shape)

(77241, 35)
(33104, 35)


#### Export Data

In [ ]:
# train export
train_3.to_csv('../소_3대 정리/3_train_utf.csv', index=False) # utf
train_3.to_csv('../소_3대 정리/3_train_cp.csv',encoding= 'cp949', index=False) # cp949
# test export 
test_3.to_csv('../소_3대 정리/3_test_utf.csv', index=False) # utf
test_3.to_csv('../소_3대 정리/3_test_cp.csv',encoding= 'cp949', index=False) # cp949

In [33]:
train_3.columns

Index(['이표번호', 'KPN', '성별', '출생일', '농가코드', '농가등급', '도축개월', '도축일', '등급', '도체중',
       '등심단면적', '등지방두께', '육색', '지방색', '성숙도', '조직도', '어미 이표번호', '형매 도체중 평균',
       '형매 등심단면적 평균', '형매 등지방 평균', '형매 근내지방 평균', '형매 마릿수', '외할미 이표번호',
       '어미형매 도체중 평균', '어미형매 등심단면적 평균', '어미형매 등지방 평균', '어미형매 근내지방 평균',
       '어미형매 마릿수', '외증조할미 이표번호', '외할미형매 도체중 평균', '외할미형매 등심단면적 평균',
       '외할미형매 등지방 평균', '외할미형매 근내지방 평균', '외할미형매 마릿수', '근내지방도'],
      dtype='object')

##### 농장 번호에 따라 소들의 근내지방평균
> 1. <font color = pink> Pivot table </font>
> 2. <font color = pink> 시각화 </font>

##### Import Data

In [5]:
full_df = pd.read_csv('3_total_cp.csv', encoding='cp949')
train = pd.read_csv('3_train_cp.csv', encoding='cp949')
test = pd.read_csv('3_test_cp.csv', encoding = 'cp949')

#### 농가 분류를 위한 작업
> 1. 농가 id 고유값 확인
> 2. 농가 id에 따른 등급이 있는데, 각 id별 farm_class가 동일한지 확인

In [6]:
print(full_df.shape)
print(f'농가_id, unique 개수:{len(full_df.farm_id.unique())}')
print('\n')
print(full_df.farm_id.value_counts())

(110345, 35)
농가_id, unique 개수:26248


407775    256
228166    188
380843    164
125221    162
254273    153
         ... 
404599      1
462766      1
204315      1
409968      1
400158      1
Name: farm_id, Length: 26248, dtype: int64


In [85]:
co_a = pd.DataFrame(full_df.farm_id.value_counts())
co_a.columns=['farm_id_count']
print(co_a.head())
print('\n')
print(len(co_a[co_a['farm_id_count']==5]))

        farm_id_count
407775            256
228166            188
380843            164
125221            162
254273            153


1119


##### Farm_id별 class를 확인했을 때, 전부 같은 등급인 것 같음
> ☞ 중복값 제거를 통해 각 id 고유값 개수인 26248개의 행만 남으면 농가 등급은 id 별로 모두 동일한 것

In [73]:
print('263150농가의 등급')
print(full_df[full_df['farm_id']==263150]['farm_class'].head())
print(full_df[full_df['farm_id']==263150]['farm_class'].tail())
print('\n')
print('407775농가의 등급')
print(full_df[full_df['farm_id']==407775]['farm_class'].head())
print(full_df[full_df['farm_id']==407775]['farm_class'].tail())

263150농가의 등급
1       1A
64      1A
2302    1A
3876    1A
4658    1A
Name: farm_class, dtype: object
99531     1A
101280    1A
103353    1A
105053    1A
106937    1A
Name: farm_class, dtype: object


407775농가의 등급
1828    1+B
2313    1+B
2340    1+B
6899    1+B
7429    1+B
Name: farm_class, dtype: object
92602    1+B
93066    1+B
93313    1+B
93327    1+B
97281    1+B
Name: farm_class, dtype: object


##### 등급 중복 확인을 위한 중복값 제거 작업

In [22]:
du_full = full_df.copy()
du_1 = du_full.drop_duplicates(['farm_id','farm_class'], keep='first') 
print(du_1.shape)
print(du_1.head())
# 결과가 애매함, 애초에 farm_id로 중복값을 제거먼저 해줬으므로 정확하다고 할 수 없음

(26248, 35)
           id      kpn gender         bir  farm_id farm_class  sl_m  \
0  2137173111  KPN1053      F  2019-03-28   398841        1+C    42   
1  2150109802  KPN1060    N_M  2020-04-28   263150         1A    28   
2  2133835356  KPN1099      F  2018-09-28   189147         1A    52   
3  2156452289  KPN1284      F  2020-11-18   451221         1A    27   
4  2147835365  KPN1060      F  2019-10-23   190620         1A    32   

         sl_d class  s_w  ...  a_s_m_i  a_s_f_m  a_s_t_m  a_s_c        b_id  \
0  2022-09-21    1B  374  ...    78.33      9.0     5.33      3  2054234523   
1  2022-08-22  1++C  524  ...   105.83     16.0     7.33      6  2019892786   
2  2023-01-03    1A  333  ...    90.50     13.0     4.00      2  2060071317   
3  2023-01-26    1A  372  ...   109.00     18.0     8.00      1  2069804608   
4  2022-06-07    2A  378  ...   109.00     18.0     8.00      1  2069804608   

   b_s_m_w  b_s_m_i  b_s_f_m  b_s_t_m  b_s_c  
0    571.0     95.0     10.0      5.0  

##### 중복 제거로 확인할 수 없으므로 피벗테이블로 최빈, 최소값 등 비교를 통해 각 id내 다른 값이 있는지 확인

In [24]:
# 농가 id별 class 최빈값
a1 = full_df.groupby('farm_id')['farm_class'].agg(**{
    'most_common_value':lambda x:x.mode()
}).reset_index()

In [25]:
a1

farm_id most_common_value
0         1571               1+B
1         1580                1A
2         1591                1A
3         1597               1+B
4         1598                1B
...        ...               ...
26243   471434               1+C
26244   471736                1B
26245   471813               1+B
26246   471837                1C
26247   517575                2B

[26248 rows x 2 columns]

In [77]:
a2 = full_df.groupby(['farm_id', 'farm_class'])[['target']].agg(['count'])

In [34]:
a2.to_csv('각id별_class확인.csv', encoding='cp949', index=False)

#### 농가 id별 target 평균

In [65]:
gr_target = full_df.groupby(['farm_id'])['target'].agg(**{'mean_target':'mean'}).reset_index()

In [66]:
# 소숫점 2자리까지 반올림
gr_target.mean_target = gr_target.mean_target.round(2)

In [67]:
# 내림차순 정렬 및 원래 농가 등급을 id별로 붙여 넣어 비교해보기

# 1. target 평균 기준으로 내림차순 정렬
gr_t = gr_target.copy()
gr_t.sort_values('mean_target', ascending=False, inplace=True)
print(gr_t.head())

# 2. target 평균 값과 비교하여 농가 등급이 진짜 우수한지, 원 데이터의 농가 등급 변수를 추가하기
farm = full_df[['farm_id', 'farm_class']].reset_index()
df_mer = pd.merge(gr_t, farm, how='left', left_on= 'farm_id', right_on='farm_id')


       farm_id  mean_target
13132   315432          1.0
18623   399342          1.0
18749   401012          1.0
18735   400761          1.0
18732   400733          1.0


In [70]:
df_mer.to_csv('소_농가별_근내지방도평균.csv',index=False, encoding='cp949')

#### 농가 등급 A, B, C으로 변수 추가

In [90]:
print(full_df.farm_class.unique())
print(full_df.farm_class.value_counts().sort_index())

['1+C' '1A' '1++C' '1C' '1+B' '2A' '1B' '1+A' '1++B' '2B' '3A' '2C' '3B'
 '1++A' '3C' '등외']
1++A       65
1++B      273
1++C      870
1+A      4220
1+B     17657
1+C     32803
1A      28067
1B      15885
1C       6804
2A       2231
2B        927
2C        306
3A         87
3B        103
3C         29
등외         18
Name: farm_class, dtype: int64


In [94]:
full_df['farm_level'] = full_df['farm_class'].apply(lambda x: 'A' if ((x == '1++A')|(x == '1++B')|(x == '1++C'))  else 'B' if ((x == '1+A')|(x == '1+B')|(x == '1+C')) else 'C')

In [98]:
a_cl = full_df[full_df['farm_level']=='A']
b_cl = full_df[full_df['farm_level']=='B']
c_cl = full_df[full_df['farm_level']=='C']
a_cl = a_cl[['farm_id','farm_class','farm_level']]
b_cl = b_cl[['farm_id','farm_class','farm_level']]
c_cl = c_cl[['farm_id','farm_class','farm_level']]

In [100]:
print(a_cl.farm_class.unique())
print(b_cl.farm_class.unique())
print(c_cl.farm_class.unique())

['1++C' '1++B' '1++A']
['1+C' '1+B' '1+A']
['1A' '1C' '2A' '1B' '2B' '3A' '2C' '3B' '3C' '등외']


In [86]:
# 결측치 확인
full_df.farm_class.isnull().sum()

0

In [105]:
full_df.to_csv('등급추가_full_cp.csv', encoding='cp949', index=False)

#### Univariate Logistic Regression

In [102]:
full_df.columns

Index(['id', 'kpn', 'gender', 'bir', 'farm_id', 'farm_class', 'sl_m', 'sl_d',
       'class', 's_w', 'i_a', 'f_t', 'target', 'b_c', 'f_c', 'm_l', 'organ',
       'mom_id', 's_m_w', 's_m_i', 's_f_m', 's_t_m', 's_c', 'a_id', 'a_s_m_w',
       'a_s_m_i', 'a_s_f_m', 'a_s_t_m', 'a_s_c', 'b_id', 'b_s_m_w', 'b_s_m_i',
       'b_s_f_m', 'b_s_t_m', 'b_s_c', 'farm_level'],
      dtype='object')

In [119]:
# 사용할 변수만 고르기
logit_df = full_df.copy()
uni_logit_df = full_df[['gender', 's_w', 'sl_m',  's_m_w', 's_m_i', 's_f_m', 's_t_m', 's_c', 'a_s_m_w',
       'a_s_m_i', 'a_s_f_m', 'a_s_t_m', 'a_s_c', 'b_s_m_w', 'b_s_m_i',
       'b_s_f_m', 'b_s_t_m', 'b_s_c', 'farm_level','target']]
print(uni_logit_df.shape)
print(uni_logit_df.columns)

(110345, 20)
Index(['gender', 's_w', 'sl_m', 's_m_w', 's_m_i', 's_f_m', 's_t_m', 's_c',
       'a_s_m_w', 'a_s_m_i', 'a_s_f_m', 'a_s_t_m', 'a_s_c', 'b_s_m_w',
       'b_s_m_i', 'b_s_f_m', 'b_s_t_m', 'b_s_c', 'farm_level', 'target'],
      dtype='object')


In [122]:
full_df.farm_level.value_counts()

B    54680
C    54457
A     1208
Name: farm_level, dtype: int64

#### 반복문을 이용하여 단변량 로지스틱 돌리기

In [121]:
for i in range(0,19):
    print(f'순서:{i+1}번째 변수')
    model = sm.Logit.from_formula('target ~ uni_logit_df.iloc[:,i]', uni_logit_df).fit()
    print(f'독립변수 이름: {uni_logit_df.columns[i]}')
    print(model.summary())
    print('============='*3)
    print(f'모델 AIC:{model.aic}')
    print('\n')
    print(f'오즈비{np.exp(model.params)}')
    print('\n')

순서:1번째 변수
Optimization terminated successfully.
         Current function value: 0.577376
         Iterations 5
독립변수 이름: gender
                           Logit Regression Results                           
Dep. Variable:                 target   No. Observations:               110171
Model:                          Logit   Df Residuals:                   110169
Method:                           MLE   Df Model:                            1
Date:                Thu, 09 Feb 2023   Pseudo R-squ.:                 0.03802
Time:                        17:52:13   Log-Likelihood:                -63610.
converged:                       True   LL-Null:                       -66124.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                         -1.6037 